aggregator/leader_flush_mgr.go

```diff
@@ -0,0 +1,371 @@
+func (mgr *leaderFlushManager) CanLead() bool { return true }
+
+func (mgr *leaderFlushManager) enqueueBucketWithLock(
+	bucketIdx int,
+	bucket *flushBucket,
+) {
+	flushInterval := bucket.interval
+	nextFlushNanos := mgr.computeNextFlushNanos(flushInterval)
+	newFlushMetadata := flushMetadata{
+		timeNanos: nextFlushNanos,
+		bucketIdx: bucketIdx,
+	}
+	mgr.flushTimes.Push(newFlushMetadata)
+}
+
+func (mgr *leaderFlushManager) computeNextFlushNanos(flushInterval time.Duration) int64 {
+	now := mgr.nowFn()
+	nextFlushNanos := now.UnixNano() + int64(flushInterval)
+	if mgr.jitterEnabled {
+		alignedNow := now.Truncate(flushInterval)
+		if alignedNow.Before(now) {
+			alignedNow = alignedNow.Add(flushInterval)
+		}
+		maxJitter := mgr.maxJitterFn(flushInterval)
+		jitterNanos := mgr.randFn(int64(maxJitter))
+		nextFlushNanos = alignedNow.UnixNano() + jitterNanos
+	}
+	return nextFlushNanos
+}
+
+func (mgr *leaderFlushManager) prepareFlushTimesWithLock(
+	buckets []*flushBucket,
+) schema.ShardSetFlushTimes {
+	proto := schema.ShardSetFlushTimes{
+		ByShard: make(map[uint32]*schema.ShardFlushTimes, defaultInitialFlushTimesCapacity),
+	}
+	for _, bucket := range buckets {
+		for _, flusher := range bucket.flushers {
+			shard := flusher.Shard()
+			shardFlushTimes, exists := proto.ByShard[shard]
+			if !exists {
+				shardFlushTimes = &schema.ShardFlushTimes{
+					ByResolution: make(map[int64]int64, 2),
+				}
+				proto.ByShard[shard] = shardFlushTimes
+			}
+			resolution := flusher.Resolution()
+			lastFlushedNanos := flusher.LastFlushedNanos()
+			shardFlushTimes.ByResolution[int64(resolution)] = lastFlushedNanos
+		}
+	}
+	return proto
+}
+
+func (mgr *leaderFlushManager) persistFlushTimes(flushTimes *schema.ShardSetFlushTimes) {
+	persistStart := mgr.nowFn()
+	persistErr := mgr.flushTimesPersistRetrier.Attempt(func() error {
+		_, err := mgr.flushTimesStore.Set(mgr.flushTimesKey, flushTimes)
+		return err
+	})
+	mgr.metrics.flushTimesPersist.ReportSuccessOrError(persistErr, mgr.nowFn().Sub(persistStart))
+}
+
+func (mgr *leaderFlushManager) nowNanos() int64 { return mgr.nowFn().UnixNano() }
+
+type leaderFlushTask struct {
+	mgr           *leaderFlushManager
+	shouldFlush   bool
+	duration      tally.Timer
+	flushers      []PeriodicFlusher
+	shouldPersist bool
+	flushTimes    *schema.ShardSetFlushTimes
+}
+
+func (t *leaderFlushTask) Run() {
+	mgr := t.mgr
+
+	// NB(xichen): we intentionally do not use the worker pool as persistence could be
+	// done asynchronously to avoid blocking the workers reserved for flushing. In theory
+	// if the persistence takes a long to finish, there is a chance two goroutines may
+	// persist flush times at the same time. However in practice this is okay because kv
+	// persistence is fast so this rarely if at all happens. And even if it happens, the
+	// impact is that the follower may flush less data than it optimally could but that
+	// does not affect the correctness of the system.
```

Comment:
nit: why not just order the kv.Set() by having a thread listen on a channel for notification, and in Run() just send a signal to the channel? I know the current logic is not incorrect. But just to be more careful here?